sdk/cosmos/azure-cosmos/src/main/java/com/azure/cosmos/CosmosAsyncContainer.java

```diff
@@ -310,8 +317,194 @@ public <T> Mono<CosmosItemResponse<T>> createItem(T item, CosmosItemRequestOptio
         return withContext(context -> createItemInternal(item, requestOptions, context));
     }
 
+    private static void mergeDiagnostics(CosmosException originalCosmosException, CosmosException readCosmosError) {
+        checkNotNull(originalCosmosException, "Argument 'originalCosmosException' must not be null.");
+        checkNotNull(readCosmosError, "Argument 'readCosmosError' must not be null.");
+
+        CosmosDiagnostics readDiagnostics = readCosmosError.getDiagnostics();
+        if (readDiagnostics != null && readDiagnostics.getClientSideRequestStatisticsRaw() != null) {
+            CosmosDiagnostics originalDiagnostics = originalCosmosException.getDiagnostics();
+            if (originalDiagnostics == null
+                || originalDiagnostics.getClientSideRequestStatisticsRaw() == null) {
+
+                originalCosmosException.setDiagnostics(readDiagnostics);
+            } else {
+                originalDiagnostics.clientSideRequestStatistics().recordContributingPointOperation(
+                    readDiagnostics.getClientSideRequestStatisticsRaw()
+                );
+            }
+        }
+    }
+
+    private static void mergeDiagnostics(
+        ResourceResponse<Document> readResponse,
+        CosmosException originalCosmosException) {
+
+        CosmosDiagnostics responseDiagnostics = readResponse.getDiagnostics();
+        if (responseDiagnostics != null &&
+            responseDiagnostics.getClientSideRequestStatisticsRaw() != null) {
+
+            CosmosDiagnostics errorDiagnostics = originalCosmosException.getDiagnostics();
+            if (errorDiagnostics != null) {
+                responseDiagnostics.clientSideRequestStatistics().recordContributingPointOperation(
+                    errorDiagnostics.getClientSideRequestStatisticsRaw()
+                );
+            }
+
+            readResponse.addRequestCharge(originalCosmosException.getRequestCharge());
+        }
+    }
+
+    private <T> Mono<CosmosItemResponse<T>> replaceItemWithRetriesInternal(Class<T> itemType,
```

Comment:
the naming here might be a little misleading - because even in `replaceItemInternalCore`, retry can happen depends on the writeRetryPolicy. The only difference here is to convert into a success when applicable.  Maybe  `replaceItemWithTrackingId` or similar?

Reply:
Good point